espree.js

```diff
@@ -66,23 +111,43 @@ import { getLatestEcmaVersion, getSupportedEcmaVersions } from "./lib/options.js
 
 // To initialize lazily.
 const parsers = {
-    _regular: null,
-    _jsx: null,
+    _regular: /** @type {IEspreeParser|null} */ (null),
+    _jsx: /** @type {IEspreeParser|null} */ (null),
 
+    /**
+     * Returns regular Parser
+     * @returns {IEspreeParser} Regular Acorn parser
+     */
     get regular() {
         if (this._regular === null) {
-            this._regular = acorn.Parser.extend(espree());
+            const espreeParserFactory = espree();
+
+            // Cast the `acorn.Parser` to our own for required properties not specified in *.d.ts
+            this._regular = espreeParserFactory(/** @type {AcornJsxParser} */ (acorn.Parser));
```

Comment:
Why can't we keep using `acorn.Parser.extend`? It's the [official way](https://github.com/acornjs/acorn/tree/master/acorn#the-parser-class) to use Acorn plugins.

Reply:
TypeScript complains that the construct signatures are not compatible. For example, `EspreeParser` allows for `null` options, Acorn requires `ecmaVersion` now, etc. While one might think that Espree's looser requirements would allow Espree to be permitted here by TS, it is not.

Reply:
> TypeScript complains that the construct signatures are not compatible.

Does this mean that the types for `acorn.Parser.extend` are wrong, or that we were using ``acorn.Parser.extend`` in a way it wasn't actually intended to be used?

Reply:
The types should be fine, and we were using it fine, but it was just easier to coerce the types.

I've added a commit to resume using the `extend` API.

Note that this change highlights one shortcoming in `lib/espree.js`.  For the export:

```js
export default () => {

    /**
     * Returns the Espree parser.
     * @param {AcornJsxParser} Parser The Acorn parser
     * @returns {typeof EspreeParser} The Espree parser
     */
    return Parser => {
      // ...
    };
```

...the `@param` should, I think, really be `typeof acorn.Parser | AcornJsxParser` since it can accept a plain Acorn parser as well as a JSX one.

If we did this, we should be able to drop the `unknown` conversion (since the plain Acorn parser would be known to be acceptable), and it seems to fit more semantically.

However, this introduces a whole can of worms in that, with a union of these two class signatures, TypeScript no longer properly finds the various Acorn base parser properties (despite the fact that either class gives them), and we need to inform TS of their existence by indicating a great deal of them explicitly in the constructor.

Another option might be to just accept `typeof acorn.Parser` as the param, but then TS complains about the JSX-specific properties being optionally used within the class, and adding duck-checking before their use doesn't appear to satisfy TS to signal that the properties do not always exist (and don't need to exist).

Or, we could just leave it as is.